lib/net/brontide.js

```diff
@@ -9,6 +9,95 @@
+ *   sender must provide the receiver with two preimages
+ *   (u, v) which map to a point via `f^2(u,v) -> f(u) + f(v)`
+ *   where `f(u)` is the forward map. The inverse map f^2^-1(p)
+ *   is more complex and involves a loop as well as random
+ *   group element generation.
+ *
+ *   Performance may be a concern. Encoding a point requires
+ *   16 inversions and 20 square roots on average (an average
+ *   of 4 iterations of the elligator squared loop). This
+ *   should be on the order of ~2-3 point multiplications.
+ *   In the future, the inversions can likely be optimized
+ *   and reduced if multiple iterations are attempted at
+ *   once.
+ *
+ *   On its face, it seems like this loop should be the
+ *   subject of timing attacks. An adversary may run the
+ *   forward map and gauge their own timing with our's.
+ *   This is not the case for one simple reason: the
+ *   timing is _random_ with elligator squared, as a
+ *   random group element is generated on each iteration.
+ *
+ *   The forward map is cheaper, requiring only 6 square
+ *   roots and 3 inversions. We can visualize this as decoding
+ *   _nine_ public keys instead of one.
+ *
+ *   @sipa proposed what I'm referring to as the "unsquared"
+ *   version of SVDW inversion[3], but this does not
+ *   necessarily save on performance as a Legendre symbol
+ *   has to be computed for every call to the forward map:
+ *   this is roughly as expensive as the square root it
+ *   replaces (furthermore, @sipa's inverse function and
+ *   distribution seem to be incorrect (?)).
+ *
+ *   We can potentially double the performance by moving to
+ *   a curve which is 3-isogenous to secp256k1[4] (idea from [5]).
+ *   This allows us to use the Simplified Shallue-Woestijne-Ulas
+ *   map instead of the Shallue-van de Woestijne map, though,
+ *   the sender may have to modify their private key before
+ *   executing the actual ECDH (as their public key would
+ *   be multiplied by 3 by the receiver).
+ *
+ *   All bets are off if the user is MITM'd, or rather _tricked_
+ *   into connecting to an adversary via their published
+ *   identity key. A honeypot attack like this will no doubt
+ *   prove that this particular exchange of messages is that
+ *   of an end-to-end encryption handshake. However, the
+ *   construction is here to provide a mitigation against
+ *   more passive attacks, like non-interactive packet
+ *   inspection, not MITMs (see also: the great firewall).
+ *
+ *   A future possibility may be to randomly pad out messages
+ *   to avoid suspicion from a tyrannical government
+ *   (todo: research/discuss this more).
+ *
+ *   We have an elligator squared implentation in both C and
+ *   javascript[6][7].
+ *
+ *   [1] https://eprint.iacr.org/2014/043.pdf
+ *   [2] https://www.di.ens.fr/~fouque/pub/latincrypt12.pdf
+ *   [3] https://gist.github.com/sipa/29118d3fcfac69f9930d57433316c039
+ *   [4] https://github.com/bcoin-org/bcrypto/blob/master/scripts/secp256k1.sage
```

Comment:
`404`... did you mean https://github.com/bcoin-org/bcrypto/blob/master/scripts/elligator.sage ?

Reply:
No, just a file I forgot to commit. Better explanation here now:

- https://gist.github.com/chjj/09c447047d5d0b63afcbbbc484d2c882
- https://github.com/bcoin-org/bcrypto/commit/aac4464c507c823ebcf62501d46d8f427aab4db2